File: openc3/lib/openc3/utilities.rb
# encoding: ascii-8bit

# Copyright 2022 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# Modified by OpenC3, Inc.
# All changes Copyright 2022, OpenC3, Inc.
# All Rights Reserved

module OpenC3
  autoload(:Logger, 'openc3/utilities/logger.rb')
  autoload(:Authorization, 'openc3/utilities/authorization.rb')
  autoload(:Store, 'openc3/utilities/store_autoload.rb')
  autoload(:Sleeper, 'openc3/utilities/sleeper.rb')
  autoload(:Crc, 'openc3/utilities/crc.rb')
  autoload(:Crc16, 'openc3/utilities/crc.rb')
  autoload(:Crc32, 'openc3/utilities/crc.rb')
  autoload(:Crc64, 'openc3/utilities/crc.rb')
  autoload(:Csv, 'openc3/utilities/csv.rb')
  autoload(:Metric, 'openc3/utilities/metric.rb')
  autoload(:MessageLog, 'openc3/utilities/message_log.rb')
  autoload(:Quaternion, 'openc3/utilities/quaternion.rb')
  autoload(:SimulatedTarget, 'openc3/utilities/simulated_target.rb')
end
autoload(:RubyLexUtils, 'openc3/utilities/ruby_lex_utils.rb')

File: openc3-init/plugins/packages/openc3-tool-scriptrunner/src/tools/ScriptRunner/autocomplete/utilities.js
/*
# Copyright 2021 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# Modified by OpenC3, Inc.
# All changes Copyright 2022, OpenC3, Inc.
# All Rights Reserved
*/

import Api from '@openc3/tool-common/src/services/api'

const getKeywords = (type) => {
  return Api.get(`/openc3-api/autocomplete/keywords/${type}`)
}

const getAutocompleteData = (type) => {
  return Api.get(`/openc3-api/autocomplete/data/${type}`)
}

// This should probably go in some higher up util library thing place
const groupBy = (array, lambda) => {
  return array.reduce((groups, item) => {
    const key = lambda(item)
    if (groups[key]) {
      groups[key].push(item)
    } else {
      groups[key] = [item]
    }
    return groups
  }, {})
}

export { getKeywords, getAutocompleteData, groupBy }
